File: 157.read_N_characters_given_read4/read.py
# The read4 API is already defined for you.
# @param buf, a list of characters
# @return an integer
# def read4(buf):

class Solution(object):
    def read(self, buf, n):
        """
        :type buf: Destination buffer (List[str])
        :type n: Maximum number of characters to read (int)
        :rtype: The number of characters read (int)
        """
        idx = 0
        while n > 0:
            buf4 = [""] * 4
            l = read4(buf4)
            if not l: # if no more char in file in 
                return idx
            for i in xrange(min(l, n)): # 
                buf[idx] = buf4[i]
                idx += 1
                n -= 1
        return idx
'''
the idea is to extend read4 api to readn, so the funciton should be the same,
buf should be a cache and the function should return the length
'''

File: 157.read_N_characters_given_read4/read.js
/**
 * Definition for read4()
 *
 * @param {character[]} buf Destination buffer
 * @return {number} The number of actual characters read
 * read4 = function(buf) {
 *     ...
 * };
 */

/**
 * @param {function} read4()
 * @return {function}
 */
var solution = function(read4) {
  /**
   * @param {character[]} buf Destination buffer
   * @param {number} n Number of characters to read
   * @return {number} The number of actual characters read
   */
  return function(buf, n) {
    let idx = 0;
    while (n > 0) {
      let buf4 = new Array(4);
      const l = read4(buf4);
      if (l == 0) return idx;
      const range = Math.min(l, n);
      // can't set Math.min(l,n) as the boundary directly in the condition expression
      // because n changes everytime
      for (let i = 0; i < range; i++) {
        buf[idx] = buf4[i];
        idx++;
        n--;
      }
    }
    return idx;
  };
};
